### PR TITLE
chore: README updates

### PR DESCRIPTION
**Pull request by le1nux**
# What does this PR do?

This PR improves and updates the README

## General Changes
* ..

## Breaking Changes
* .. 

## Checklist before submitting final PR
- [ ] My PR is minimal and addresses one issue in isolation
- [ ] I have merged the latest version of the target branch into this feature branch
- [ ] I have reviewed my own code w.r.t. correct implementation, missing type hints, proper documentation, etc.
- [ ] I have run a sample config for model training
- [ ] I have checked that all tests run through (`python tests/tests.py`)
- [ ] I have updated the internal changelog (`CHANGELOG_DEV.md`)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -16,34 +16,78 @@
   </a>
 </div>
 
+Modalities is a pytorch-native framework for distributed training of Large Language Models (LLMs) and Foundation Models (FMs) at scale. Given the complexity of distributed training and rapid advancements in the field, we aim to provide a flexible and easy-to-use framework that enables researchers and practitioners to train and evaluate LLMs and FMs efficiently. Modalities is built on top of PyTorch and leverages the latest advancements in distributed training, such as Fully Sharded Data Parallel (FSDP), mixed precision training, Flash Attention and many more, to achieve state-of-the-art performance and throughput.
```
Comment (mali-git) [inline README.md:19]:
```suggestion
Modalities is a PyTorch-native framework for distributed training of Large Language Models (LLMs) and Foundation Models (FMs) at scale. Given the complexity of distributed training and rapid advancements in the field, we aim to provide a flexible and easy-to-use framework that enables researchers and practitioners to train and evaluate LLMs and FMs efficiently. Modalities is built on top of PyTorch and leverages the latest advancements in distributed training, such as Fully Sharded Data Parallel (FSDP), mixed precision training, Flash Attention and many more, to achieve state-of-the-art performance and throughput.
```

File: README.md [hunk 2]
```diff
@@ -16,34 +16,78 @@
...
 
+We successfully scaled modalities up tp 2048 GPUs on two HPC centers, namely [Leonardo Booster](https://leonardo-supercomputer.cineca.eu/hpc-system/) and [MareNostrum 5](https://www.bsc.es/ca/marenostrum/marenostrum-5), featuring Nvidia A100 and H100 GPUs, respectively. The results of our scaling experiments can be found [here](#scaling-experiments).
```
Comment (mali-git) [inline README.md:21]:
```suggestion
We successfully scaled Modalities up to 2048 GPUs on two HPC centers, namely [Leonardo Booster](https://leonardo-supercomputer.cineca.eu/hpc-system/) and [MareNostrum 5](https://www.bsc.es/ca/marenostrum/marenostrum-5), featuring Nvidia A100 and H100 GPUs, respectively. The results of our scaling experiments can be found [here](#scaling-experiments).
```

File: README.md [hunk 3]
```diff
@@ -16,34 +16,78 @@
...
 
+Besides its scalabilty, Modalities allows to seamlessly integrate new components and features, such as custom attention mechanisms, loss functions, optimizers or models. We provide a series of tutorials to help you get started with training and evaluating models using modalities. We achieve this level of extensibility by having clear interfaces for each component type (e.g., model, optimizer, etc.), that a component must implement to be registered within in modalities at runtime.
```
Comment (mali-git) [inline README.md:23]:
```suggestion
Besides its scalabilty, Modalities allows to seamlessly integrate new components and features, such as custom attention mechanisms, loss functions, optimizers or models. We provide a series of tutorials to help you get started with training and evaluating models using Modalities. We achieve this level of extensibility by having clear interfaces for each component type (e.g., model, optimizer, etc.), that a component must implement to be registered within in Modalities at runtime. 
```

File: README.md [hunk 4]
```diff
@@ -16,34 +16,78 @@
...
 ## Getting Started
 For training and evaluation a model, feel free to checkout [this](https://github.com/Modalities/modalities/blob/main/examples/getting_started/README.md) getting started tutorial, in which we train a small, 60M-parameter GPT model on a tiny subset of the Redpajama V2 dataset. 
 
 ## Installation
 
-Create conda environment and activate it via 
-```
+There are two ways to install modalities. If you want to use the latest nightly version, or if you want to modify the code base itself, we recommend installing modalities directly from source. 
+
+If you want to use modalities as a library and potentially register your custom components with modalities, you can install it directly via pip which provides you with the latest stable version.
```
Comment (mali-git) [inline README.md:32]:
```suggestion
If you want to use Modalities as a library and register your custom components with Modalities, you can install it directly via pip which provides you with the latest stable version.
```

File: README.md [hunk 5]
```diff
@@ -16,34 +16,78 @@
...
 
+Currently, the flash attention dependency cannot be installed without torch being installed beforehand.
+Until the flash attention developers fix this, we have to run
+
+```sh
+pip install torch
 ```
+beforehand.
+
+Afterwards, modalities can be installed via
```
Comment (mali-git) [inline README.md:63]:
```suggestion
Afterwards, Modalities can be installed via
```

File: README.md [hunk 6]
```diff
@@ -16,34 +16,78 @@
...
+### Option 2: Installation via pip
 
+To install modalities via pip, run
+
+```sh
+pip install torch
+pip install modalities
+```
+
+Note, that also here, torch has to be installed before installing modalities due to flash attention's dependency management.
```
Comment (mali-git) [inline README.md:79]:
```suggestion
Note, that also here, torch has to be installed before installing Modalities due to flash attention's dependency management.
```

File: README.md [hunk 7]
```diff
@@ -16,34 +16,78 @@
...
+For running the training endpoint on multiple GPUs run 
+```sh 
+CUDA_VISIBLE_DEVICES=2,3 torchrun --nnodes 1 --nproc_per_node 2 --rdzv-endpoint=0.0.0.0:29502 modalities run --config_file_path config_files/config.yaml
+```
 
-Or, if you are a VsCode user, add this to your `launch.json`:
+In the example above, we use `torchrun` to run the training endpoint on two GPUs. The `--nnodes` argument specifies the number of nodes in the cluster, `--nproc_per_node` specifies the number of processes per node, and `--rdzv-endpoint` specifies the rendezvous endpoint. The `modalities run` command specifies the training endpoint, and `--config_file_path` specifies the path to the configuration file. The configuraton file contains the exhaustive parameterization for all the training components (e.g., dataset, model, optimize, etc.), making training fully reproducible. A full list of all the components already available in modalities an be found [here](docs/components/components.md).
```
Comment (mali-git) [inline README.md:88]:
```suggestion
In the example above, we use `torchrun` to run the training endpoint on two GPUs. The `--nnodes` argument specifies the number of nodes in the cluster, `--nproc_per_node` specifies the number of processes per node, and `--rdzv-endpoint` specifies the rendezvous endpoint. The `modalities run` command specifies the training endpoint, and `--config_file_path` specifies the path to the configuration file. The configuraton file contains the exhaustive parameterization for all the training components (e.g., dataset, model, optimize, etc.), making training fully reproducible. A full list of all the components already available in Modalities can be found [here](docs/components/components.md).
```

File: README.md [hunk 8]
```diff
@@ -107,60 +157,83 @@ In the following, we list the already implemented, planned and in-progress featu
 | Knowledge Distillation         | planned  | Transfers knowledge from a larger, complex model to a smaller, more efficient model, improving the smaller model's performance without the computational cost of the larger model.|
 | Hyperparameter Optimization    | planned          | Grid search for various hyperparameter such as LR, Optimizer arguments etc. Also the integration of µP might be interesting |
 
+## Tutorials
+Even though modalities significantly simplifies LLM training, there is still some technical complexity left. We provide a series of tutorials to help you get started with training and evaluating models using modalities.
```
Comment (mali-git) [inline README.md:161]:
```suggestion
Even though Modalities significantly simplifies LLM training, there is still some technical complexity left. We provide a series of tutorials to help you get started with training and evaluating models using Modalities.
```

File: README.md [hunk 9]
```diff
@@ -107,60 +157,83 @@ In the following, we list the already implemented, planned and in-progress featu
...
+
+- [Getting Started](examples/getting_started/README.md)</br>
+  Brief overview on how to get started with modalities by training a small GPT model on a tiny subset of the Redpajama V2 dataset.
```
Comment (mali-git) [inline README.md:164]:
```suggestion
  Brief overview on how to get started with Modalities by training a small GPT model on a tiny subset of the Redpajama V2 dataset.
```

File: README.md [hunk 10]
```diff
@@ -107,60 +157,83 @@ In the following, we list the already implemented, planned and in-progress featu
...
+
+- [Library Usage](examples/library_usage/README.md)</br>
+  How to use modalities as a library and register custom components with modalities.
```
Comment (mali-git) [inline README.md:167]:
```suggestion
  How to use Modalities as a library and register custom components with Modalities.
```

File: README.md [hunk 11]
```diff
@@ -107,60 +157,83 @@ In the following, we list the already implemented, planned and in-progress featu
...
 
 ## Entry Points
+Modalities provides several entry points to interact with the framework. The following section lists the available entry points and their respective functionalities.
+
+
+### Model training
 
-We use [click](https://click.palletsprojects.com/en/) as a tool to add new entry points and their CLI arguments.
-For this we have a main entry point from which all other entry points are started. 
+For model pretraining, we have to pass a configuration file that specifies the model architecture, optimizer, dataset, dataloader, and other training components. Additionall, we specify the number of nodes, the number of processes per node, and the rendezvous endpoint.
```
Comment (mali-git) [inline README.md:178]:
```suggestion
For model pretraining, we have to pass a configuration file that specifies the model architecture, optimizer, dataset, dataloader, and other training components. Additionally, we specify the number of nodes, the number of processes per node, and the rendezvous endpoint. 
```

File: README.md [hunk 12]
```diff
@@ -107,60 +157,83 @@ In the following, we list the already implemented, planned and in-progress featu
...
-@click.option(
-    "--my_cli_argument",
-    type=int,
-    required=True,
-    help="New integer argument",
-)
-def entry_point_do_stuff(config_file_path: Path, my_cli_argument: int):
-    print(f"Do stuff with {config_file_path} and {my_cli_argument}...)
-    ...
+* `$(which modalities) run`: This part dynamically finds the path to the modalities executable and runs it. The run command triggers the main process to start the training.
```
Comment (mali-git) [inline README.md:200]:
```suggestion
* `$(which modalities) run`: This part dynamically finds the path to the Modalities executable and runs it. The run command triggers the main process to start the training.
```

File: examples/getting_started/README.md [hunk 13]
```diff
@@ -146,7 +146,7 @@ first and then divides it into chunks of size context-length.
 ## 2. Training & Evaluation
 
 ### Config File
-In modalities, we describe the entire training and evaluation setup (i.e., components such as model, trainer, evaluator, dataloder etc.) within a single config file. Not only does this increase reproducibility but also allows for having the entire training runs under version control. 
+In modalities, we describe the entire training and evaluation setup (i.e., components such as model, trainer, evaluator, dataloder etc.) within a single config file. Not only does this increase reproducibility but also allows for having the entire training runs under version control. A full list of all the components already available in modalities an be found [here](../../docs/components/components.md).
```
Comment (mali-git) [inline examples/getting_started/README.md:149]:
```suggestion
In Modalities, we describe the entire training and evaluation setup (i.e., components such as model, trainer, evaluator, dataloder etc.) within a single configuration file. Not only does this increase reproducibility but also allows for having the entire training runs under version control. A full list of all the components already available in modalities an be found [here](../../docs/components/components.md).
```

File: examples/library_usage/README.md [hunk 14]
```diff
@@ -1,6 +1,6 @@
...
+Modalities can be used in a library fashion by installing the package via `pip`, as described in the [README](https://github.com/Modalities/modalities?tab=readme-ov-file#installation). The framework allows for the addition of custom components to the registry at runtime without necessitating any code changes to modalities. This functionality is achieved in Modalities with the introduction of a component registry, containing all the internal components (e.g., Dataloader, Loss function etc.). To support the addition of custom components (e.g., new model architectures) at runtime, Modalities exposes a function endpoint adding custom components to the internal registry. A full list of all the components already available in modalities an be found [here](../../docs/components/components.md).
```
Comment (mali-git) [inline examples/library_usage/README.md:3]:
```suggestion
Modalities can be used in a library fashion by installing the package via `pip`, as described in the [README](https://github.com/Modalities/modalities?tab=readme-ov-file#installation). The framework allows for the addition of custom components to the registry at runtime without necessitating any code changes to modalities. This functionality is achieved in Modalities with the introduction of a component registry, containing all the internal components (e.g., Dataloader, Loss function etc.). To support the addition of custom components (e.g., new model architectures) at runtime, Modalities exposes a function endpoint adding custom components to the internal registry. A full list of all the components already available in Modalities can be found [here](../../docs/components/components.md).
```

**Review by mali-git — CHANGES_REQUESTED** [review]

**Review by mali-git — APPROVED** [review]
